### ISSUE TITLE
Get-InstalledModule returns no results when MyDocuments is in the OneDrive folder.

### ISSUE THREAD

**Issue by aetos382** [PowerShell/PowerShellGet#509]
Steps to reproduce
------------------
Configure the OneDrive client to back up the Document directory containing your PowerShell directory to OneDrive.

```powershell
Get-InstalledModule
```

Expected behavior
-----------------

```none
Version              Name                                Repository           Description
-------              ----                                ----------           -----------
1.4.2                PackageManagement                   PSGallery            PackageManagement (a.k.a. OneGet) ...
2.1.5                PowerShellGet                       PSGallery            PowerShell module with commands f ...
```

Actual behavior
---------------

```none
(No results)
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.2.1
PSEdition                      Core
GitCommitId                    6.2.1
OS                             Microsoft Windows 10.0.18932
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

```powershell
> Get-Module -ListAvailable PowerShellGet,PackageManagement


    Directory: C:\Users\aetos\OneDrive\ドキュメント\PowerShell\Modules

ModuleType Version    Name                                PSEdition ExportedCommands
---------- -------    ----                                --------- ----------------
Script     1.4.2      PackageManagement                   Desk      {Find-Package, Get-Package, Get-PackageProvider, Get-PackageSource…}
Script     2.1.5      PowerShellGet                       Desk      {Find-Command, Find-DSCResource, Find-Module, Find-RoleCapability…}

    Directory: C:\program files\powershell\6\Modules

ModuleType Version    Name                                PSEdition ExportedCommands
---------- -------    ----                                --------- ----------------
Script     1.3.2      PackageManagement                   Desk      {Find-Package, Get-Package, Get-PackageProvider, Get-PackageSource…}
Script     2.1.3      PowerShellGet                       Desk      {Find-Command, Find-DSCResource, Find-Module, Find-RoleCapability…}
```

```powershell
> Get-PackageProvider -ListAvailable
Name                     Version          DynamicOptions
----                     -------          --------------
NuGet                    3.0.0.1          Destination, ExcludeVersion, Scope, SkipDependencies, Headers, FilterOnTag, Contains, AllowPrereleaseVersions, ConfigFile, SkipValidate
PowerShellGet            2.1.5.0          PackageManagementProvider, Type, Scope, AllowClobber, SkipPublisherCheck, InstallUpdate, NoPathUpdate, AllowPrereleaseVersions, Filter, Tag, Includes, DscResource, RoleCapability, Command, Acce…
PowerShellGet            2.1.3.0
```

OneDrive version 2019 (19.086.0502.0006)

Cause of problem
----------------
Directories in OneDrive are reparse point.
Get-InstalledModule command uses Get-ChildItem command to search for PSGetModuleInfo.xml, but can not search within the reparse point because the FollowSymLink parameter is not set.
https://github.com/PowerShell/PowerShellGet/blob/867b4f57571c36b6b82584508c9cdbd362d23715/src/PowerShellGet/private/functions/Set-InstalledModulesVariable.ps1#L10-L25

**Comment by aetos382** [comment]
Currently, Windows 10 recommends placing the Documents folder on OneDrive during its setup.
Therefore, the possibility of this problem occurring is very high.

**Comment by SydneyhSmith** [comment]
@aetos382 thanks for reporting this issue--this is likely a result of Get-InstalledModule only looking for modules which were installed using PowerShellGet and therefore only checking those default installation paths...we should address how to make this more configurable especially with so many users saving to onedrive

**Comment by kilasuit** [comment]
I have some weird results from this (though my machine probs needs a rebuild anyway)

5.1 - shows up all ok for OneDrive location
6.2 - nothing shows for OneDrive
7.0 - some modules show from the OneDrive Location, some don't

I think there's been many changes to how v6+ actually works out PSModulePath as per this list of PR's https://github.com/PowerShell/PowerShell/pulls?q=is%3Apr+PSModulePath+is%3Aclosed 

Whether this is a PowerShell or PowerShellGet issue I am not sure but I think I may need to just spin up a test VM to see if the results are any different

**Comment by csandfeld** [comment]
I ran into a similar issue with WindowsPowerShell 5.1 on two different Windows 10 installs (one is 1809 the other 1903). Both machines have MyDocuments redirected to OneDrive for business, and I found that a module installed on one machine would not show up in Get-InstalledModule on the other machine, and vice versa.

Troubleshooting the issue I found that Get-InstalledModule expects the 'hidden' attribute to be set on the modules PSGetModuleInfo.xml file, and OneDrive does not sync that attribute. After Setting the 'hidden' attribute Get-InstalledModule lists the module as expected on both machines.

Not sure if this relates to the reported problem or not, but figured I would share my finding just in case.

**Comment by glucas** [comment]
I'm seeing the same with pwsh 6.2.3, Windows 10 with my Documents folder redirected to OneDrive for  business. 

Get-Module lists several modules I've installed using PowerShellGet while Get-InstalledModule shows nothing. 

 Per the above comments:
 - PSGetModuleInfo.xml files all have the hidden attribute set, so does not seem a factor.
 - Modules were installed normally using Install-Module. My env:PSModulePath (which I did not explicitly set) has the OneDrive\Documents\PowerShell\Modules directory.

**Comment by ryanspletzer** [comment]
This have something to do with the PSModulePath and what is included in it across versions of PowerShell.

I noticed this because (using latest PowerShellGet 2.2.3 on Windows PowerShell 5.1 here), if I try to do an Update-Module -Name <moduleName> -Force for something that's in Program Files to get a new version, it by default sticks the new version of the module in my ~\OneDrive - <company>\Documents folder. However, when I subsequently go to Get-Module for the newly downloaded module it doesn't come up, and I believe this is why -- PSModulePath only includes at the ootb ~\Documents folder:

![image](https://user-images.githubusercontent.com/1481108/76686776-54f52900-65f4-11ea-80bd-7aebf8230cb3.png)

Latest PowerShellGet module (2.2.3) seems to understand where the mapped Documents folder is at in OneDrive and puts it there, but Get-Module isn't part of that module and is a builtin PowerShell cmdlet and references the PSModulePath, where it is not listed / still pointed at the normal Documents path.

I just did choco install powershell-core on this same machine for PowerShell 7 (core) and pwsh.exe takes more paths into account for PSModulePath, including the ~\OneDrive - <company>\Documents on top of ~\Documents, and therefore does find these modules.

![image](https://user-images.githubusercontent.com/1481108/76687049-c635db80-65f6-11ea-9635-1240ce25f45a.png)

Not a huge deal for 5.1 as the workaround is to just add it to PSModulePath via PS Profile. (And this is mostly a concern just on client, on server we'll almost always install to Program Files and if they're ephemeral we're rarely doing an Update-Module there.)

**Comment by aetos382** [comment]
In PowerShell 7, folders that are synced to OneDrive are specially supported, unlike regular symbolic links, so it seems to enumerate modules correctly without using the Get-ChildItem -FollowSymLink parameter.

**Comment by aetos382** [comment]
I can easily change the location of MyDocuments, but I guess PowerShellGet assumes that the path to the module directory does not change (the path when the module was installed is stored in PSGetModuleInfo.xml, It doesn't change if I change the location of MyDocuments).

What if MyDocuments on multiple computers logged in with different account names are linked to the same OneDrive account?

I think PowerShell modules isn't a document and doesn't seem to need to be backed up on OneDrive, so the module directory should be outside of MyDocuments.
A similar proposal had been submitted in the past, but was not adopted.

https://github.com/PowerShell/PowerShell/issues/8069

**Comment by ryanspletzer** [comment]
Agreed I don't think there's a reason for modules to be in OneDrive.

**Comment by kilasuit** [comment]
> Agreed I don't think there's a reason for modules to be in OneDrive.

For those that have many devices it massively simplifies switching between devices or after any device rebuilds it is a lifesaver to getting back 

I believe that the entry point that PSModule paths will use  for things like Document Location depends on what's set in the registry (on windows) 

Until user profiles as a whole can me moved eaiser between devices I do think keeping them in Documents makes sense though with work on this https://github.com/microsoft/vscode/issues/2743 expands out from just being vscode settings to generic user settings

**Comment by ryanspletzer** [comment]
Good point. I should re-phrase: I don't have a reason for modules to be in OneDrive. But I shouldn't discount others' use cases. :)

For me I have dev machine setup script(s) that help to bootstrap a lot of commonly used modules I want onto a machine (if I'm starting from scratch), and beyond that generally like to have more control when a module goes onto a given machine (installing something new to play around with) and not have it appear on all machines.

I guess that's really what it comes down to for me: it's fine for PowerShellGet to have this default, but it would be nice to have the control in PowerShellGet to override / change this default where you want your CurrentUser modules to be installed. I think we can control PSModulePath in env vars for PowerShell to determine where modules are *loaded* from, but just not where they're installed *to* for the CurrentUser scope by PowerShellGet.

**Comment by aetos382** [comment]
Users can place modules anywhere by setting $env:PSModulePath.
Users who want to put modules outside of OneDrive can do so.

But, PowerShellGet has only two options: MyDocuments or ProgramFiles.
We should be able to choose where to install modules.

Such a request I should be submitted as a proposal for PowerShellGet vNext.
